File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/irpf/LegendaDaFormulaDoIrpf.java
/*
 * Decompiled with CFR 0.152.
 */
package br.jus.trt8.pjecalc.negocio.dominio.calculo.irpf;

import br.jus.trt8.pjecalc.base.comum.HelperDate;
import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.irpf.Irpf;
import br.jus.trt8.pjecalc.negocio.dominio.ocorrenciaverba.OcorrenciaDeVerba;
import br.jus.trt8.pjecalc.negocio.dominio.verbacalculo.VerbaDeCalculo;
import java.io.Serializable;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

public class LegendaDaFormulaDoIrpf
implements Serializable {
    public static final int INDICE_ANOS_ANTERIORES = 1;
    public static final int INDICE_SEPARADO = 2;
    public static final int INDICE_EXCLUSIVA = 3;
    public static final int INDICE_NORMAL = 4;
    private static final long serialVersionUID = -7779898373287857594L;
    private Irpf irpf;
    private List<String> basesAnosAnteriores;
    private List<String> basesExclusiva;
    private List<String> basesSeparado;
    private List<String> basesNormal;

    public LegendaDaFormulaDoIrpf(Irpf irpf) {
        this.irpf = irpf;
        this.basesAnosAnteriores = new ArrayList<String>();
        this.basesExclusiva = new ArrayList<String>();
        this.basesSeparado = new ArrayList<String>();
        this.basesNormal = new ArrayList<String>();
        this.iniciar();
    }

    private void iniciar() {
        HelperDate dataLimiteAnosAnteriores = HelperDate.getInstance(this.irpf.getCalculo().getDataDeLiquidacao());
        dataLimiteAnosAnteriores.setMonth(0);
        dataLimiteAnosAnteriores.setDay(1);
        for (VerbaDeCalculo verba : this.irpf.getCalculo().getVerbas()) {
            if (!verba.getAtivo().booleanValue() || !verba.getIncidenciaIRPF().booleanValue()) continue;
            for (OcorrenciaDeVerba ocorrenciaVerbas : verba.getOcorrenciasAtivas()) {
                if (this.identificarTipoBaseCorreta(dataLimiteAnosAnteriores, verba, ocorrenciaVerbas)) break;
            }
        }
    }

    private boolean identificarTipoBaseCorreta(HelperDate dataLimiteAnosAnteriores, VerbaDeCalculo verba, OcorrenciaDeVerba ocorrenciaVerbas) {
        boolean identificou = false;
        BigDecimal valorOcorrenciaCorrigida = Utils.arredondarValorMonetario(ocorrenciaVerbas.getDiferencaCorrigida());
        if (!this.basesAnosAnteriores.contains(verba.getNome()) && HelperDate.dateBefore(ocorrenciaVerbas.getDataInicial(), dataLimiteAnosAnteriores.getDate())) {
            BigDecimal base;
            if (BigDecimal.ZERO.compareTo(valorOcorrenciaCorrigida) != 0 && Utils.naoNulo(base = ocorrenciaVerbas.getDiferencaCorrigidaParaCalculoDasIncidencias())) {
                this.basesAnosAnteriores.add(verba.getNome());
            }
        } else {
            switch (verba.getCaracteristica()) {
                case DECIMO_TERCEIRO_SALARIO: {
                    this.tratarCasoDecimoTerceiro(verba.getNome());
                    break;
                }
                case FERIAS: {
                    this.tratarCasoFerias(verba.getNome());
                    break;
                }
                case AVISO_PREVIO: 
                case COMUM: {
                    this.basesNormal.add(verba.getNome());
                }
            }
            identificou = true;
        }
        return identificou;
    }

    private void tratarCasoDecimoTerceiro(String nomeDaVerba) {
        if (this.irpf.getConsiderarTributacaoExclusiva().booleanValue()) {
            this.basesExclusiva.add(nomeDaVerba);
        } else {
            this.basesNormal.add(nomeDaVerba);
        }
    }

    private void tratarCasoFerias(String nomeDaVerba) {
        if (this.irpf.getConsiderarTributacaoEmSeparado().booleanValue()) {
            this.basesSeparado.add(nomeDaVerba);
        } else {
            this.basesNormal.add(nomeDaVerba);
        }
    }

    public List<String> getBasesAnosAnteriores() {
        return this.basesAnosAnteriores;
    }

    public List<String> getBasesSeparado() {
        return this.basesSeparado;
    }

    public List<String> getBasesExclusiva() {
        return this.basesExclusiva;
    }

    public List<String> getBasesNormal() {
        return this.basesNormal;
    }

    public String getLegenda(int tipo) {
        List<String> bases;
        StringBuilder str = new StringBuilder();
        switch (tipo) {
            case 1: {
                bases = this.basesAnosAnteriores;
                break;
            }
            case 2: {
                bases = this.basesSeparado;
                break;
            }
            case 3: {
                bases = this.basesExclusiva;
                break;
            }
            default: {
                bases = this.basesNormal;
            }
        }
        if (!bases.isEmpty()) {
            for (String nomeDaVerba : bases) {
                if (bases.indexOf(nomeDaVerba) == 0) {
                    str.append(nomeDaVerba);
                    continue;
                }
                str.append(" + ").append(nomeDaVerba);
            }
        }
        return str.toString();
    }
}

File: src/lib/pjecalc/core/dominio/calculo/irpf/legenda-da-formula-do-irpf.ts
/**
 * PJe-Calc v2.15.1 — LegendaDaFormulaDoIrpf
 * Porte 1:1 de: br.jus.trt8.pjecalc.negocio.dominio.calculo.irpf.LegendaDaFormulaDoIrpf
 *
 * Ref Java: pjecalc-fonte/.../calculo/irpf/LegendaDaFormulaDoIrpf.java (~142 linhas)
 *
 * Monta 4 listas de nomes de verbas agrupadas por regime de tributação:
 *   - basesAnosAnteriores — verbas de competência anterior a 01/01 da liquidação
 *   - basesExclusiva      — 13º quando irpf.considerarTributacaoExclusiva
 *   - basesSeparado       — férias quando irpf.considerarTributacaoEmSeparado
 *   - basesNormal         — demais (AVISO_PREVIO, COMUM, 13º/férias sem flag)
 *
 * getLegenda(tipo) concatena os nomes com " + ".
 */
import { HelperDate } from '../../../base/comum/helper-date';
import { arredondarValorMonetario } from '../../../base/comum/utils';
import type { Irpf } from './irpf';

export const INDICE_ANOS_ANTERIORES = 1;
export const INDICE_SEPARADO = 2;
export const INDICE_EXCLUSIVA = 3;
export const INDICE_NORMAL = 4;

interface VerbaLike {
  getAtivo?(): boolean;
  getIncidenciaIRPF?(): boolean;
  getNome?(): string;
  getCaracteristica?(): string;
  getOcorrenciasAtivas?(): Iterable<OcorrenciaLike>;
}
interface OcorrenciaLike {
  getDataInicial?(): Date | null;
  getDiferencaCorrigida?(): ReturnType<typeof arredondarValorMonetario>;
  getDiferencaCorrigidaParaCalculoDasIncidencias?(): ReturnType<typeof arredondarValorMonetario> | null;
}

export class LegendaDaFormulaDoIrpf {
  private irpf: Irpf;
  private basesAnosAnteriores: string[] = [];
  private basesExclusiva: string[] = [];
  private basesSeparado: string[] = [];
  private basesNormal: string[] = [];

  constructor(irpf: Irpf) {
    this.irpf = irpf;
    this.iniciar();
  }

  private iniciar(): void {
    const calc = this.irpf.getCalculo();
    if (!calc) return;
    const calcExt = calc as unknown as {
      getDataDeLiquidacao?(): Date;
      getVerbas?(): Iterable<VerbaLike>;
    };
    const dataLiq = calcExt.getDataDeLiquidacao?.();
    if (!dataLiq) return;
    const dataLimiteAnosAnteriores = new Date(dataLiq.getFullYear(), 0, 1);

    for (const verba of calcExt.getVerbas?.() ?? []) {
      if (!verba.getAtivo?.()) continue;
      if (!verba.getIncidenciaIRPF?.()) continue;
      for (const oc of verba.getOcorrenciasAtivas?.() ?? []) {
        if (this.identificarTipoBaseCorreta(dataLimiteAnosAnteriores, verba, oc)) break;
      }
    }
  }

  private identificarTipoBaseCorreta(
    dataLimiteAnosAnteriores: Date,
    verba: VerbaLike,
    oc: OcorrenciaLike,
  ): boolean {
    let identificou = false;
    const nome = verba.getNome?.() ?? '';
    const valorOcorrencia = arredondarValorMonetario(oc.getDiferencaCorrigida?.() ?? null);
    const dataInicial = oc.getDataInicial?.();
    if (!this.basesAnosAnteriores.includes(nome) && dataInicial && HelperDate.dateBefore(dataInicial, dataLimiteAnosAnteriores)) {
      const base = oc.getDiferencaCorrigidaParaCalculoDasIncidencias?.() ?? null;
      if (!valorOcorrencia.isZero() && base !== null) {
        this.basesAnosAnteriores.push(nome);
      }
    } else {
      switch (verba.getCaracteristica?.()) {
        case 'DECIMO_TERCEIRO_SALARIO':
          this.tratarCasoDecimoTerceiro(nome);
          break;
        case 'FERIAS':
          this.tratarCasoFerias(nome);
          break;
        case 'AVISO_PREVIO':
        case 'COMUM':
          this.basesNormal.push(nome);
          break;
      }
      identificou = true;
    }
    return identificou;
  }

  private tratarCasoDecimoTerceiro(nome: string): void {
    if (this.irpf.getConsiderarTributacaoExclusiva()) {
      this.basesExclusiva.push(nome);
    } else {
      this.basesNormal.push(nome);
    }
  }

  private tratarCasoFerias(nome: string): void {
    if (this.irpf.getConsiderarTributacaoEmSeparado()) {
      this.basesSeparado.push(nome);
    } else {
      this.basesNormal.push(nome);
    }
  }

  getBasesAnosAnteriores(): string[] { return this.basesAnosAnteriores; }
  getBasesSeparado(): string[] { return this.basesSeparado; }
  getBasesExclusiva(): string[] { return this.basesExclusiva; }
  getBasesNormal(): string[] { return this.basesNormal; }

  /** getLegenda (Java linha 110) — concatena por " + ". */
  getLegenda(tipo: number): string {
    let bases: string[];
    switch (tipo) {
      case INDICE_ANOS_ANTERIORES: bases = this.basesAnosAnteriores; break;
      case INDICE_SEPARADO: bases = this.basesSeparado; break;
      case INDICE_EXCLUSIVA: bases = this.basesExclusiva; break;
      default: bases = this.basesNormal;
    }
    return bases.join(' + ');
  }
}
